src/devenv/cli.py

```diff
@@ -0,0 +1,780 @@
+  version = "{get_version()}";
+  system = "{system}";
+  devenv_root = "{os.getcwd()}";
+            """)
+            # replace __DEVENV_VARS__ in flake using regex
+            flake = re.sub(r'__DEVENV_VARS__', devenv_vars, flake)
+            f.write(flake)
+
+
+@cli.command(
+    help="Deletes previous devenv generations. See http://devenv.sh/garbage-collection",
+    short_help="Deletes previous devenv generations. See http://devenv.sh/garbage-collection",
+)
+@click.pass_context
+def gc(ctx):
+    GC_ROOTS = ctx.obj['gc_root']
+    start = time.time()
+
+    # remove dangling symlinks
+    with log_task(f'Removing non-existings symlinks in {GC_ROOTS} ...'):
+        to_gc, removed_symlinks = cleanup_symlinks(GC_ROOTS)
+
+    click.echo(f'  Found {len(to_gc)} active symlinks.')
+    click.echo(f'  Deleted {len(removed_symlinks)} dangling symlinks.')
+    click.echo()
+
+    log('Running garbage collection (this process may take some time) ...', level="info")
+    # TODO: ideally nix would report some statistics about the GC as JSON
+    run_nix(f'store delete --recursive {" ".join(to_gc)}')
+
+    after_gc, removed_symlinks = cleanup_symlinks(GC_ROOTS)
+    end = time.time()
+
+    click.echo()
+    log(f'Done. Successfully removed {len(to_gc) - len(after_gc)} symlinks in {end - start:.0f} seconds.', level="info")
+
+def cleanup_symlinks(folder):
+    to_gc = []
+    removed_symlinks = []
+    for root, dirs, files in os.walk(folder):
+        for name in files:
+            full_path = os.path.join(root, name)
+            if os.path.islink(full_path):
+                if not os.path.isfile(full_path):
+                    os.unlink(full_path)
+                    removed_symlinks.append(full_path)
+                else:
+                    to_gc.append(full_path)
+    return to_gc, removed_symlinks
+
+def get_dev_environment(ctx, json=False, logging=True):
+    ctx.invoke(assemble)
+    if logging:
+        action = log_task('Building shell')
+    else:
+        action = suppress()
+    with action:
+        gc_root = os.path.join(os.environ['DEVENV_GC'], 'shell')
+        cmd = f"print-dev-env --profile '{gc_root}'"
+        if json:
+            cmd += " --json"
+        env = run_nix(cmd, logging=False, use_cachix=True)
+        run_command(f"nix-env -p '{gc_root}' --delete-generations old", logging=False, disable_stderr=True)
+        symlink_force(Path(f'{ctx.obj["gc_project"]}-shell'), gc_root)
+    return env, gc_root
+
+
+
+@cli.command(
+    help="Activate the developer environment.",
+    short_help="Activate the developer environment.",
+    context_settings=dict(
+        ignore_unknown_options=True,
+    )
+)
+@click.argument('extra_args', nargs=-1, type=click.UNPROCESSED)
+@click.argument('cmd', required=False)
+@click.pass_context
+def shell(ctx, cmd, extra_args):
+    env, gc_root = get_dev_environment(ctx)
+    if cmd:
+        run_nix(f"develop '{gc_root}' -c {cmd} {' '.join(extra_args)}", replace_shell=True)
```

Comment:
I haven't used `click` before, but currently it is not possible to pass along multiple arguments to the shell cmd:

```
$ devenv shell echo hello
...
/tmp/nix-shell.3CIEDl: line 2055: exec: hello: not found
```

The error becomes more confusing when using arguments as well:

```
$ devenv shell ls -l
...
-ls: unknown program ‘-ls’
Try '-ls --help' for more information.
```

I haven't dug into the code yet, but thought I'd share already. Will try to look into it at a later time.

I'm thinking `extra_args` might be interpreted _before_ `cmd`?

Reply:
Thanks, fixed (and added a test).